File: AES-GCM.js
const { createHash , } = require( "node:crypto" );
const { createDecipheriv , } = require( "node:crypto" );
const { randomBytes , } = require( "node:crypto" );
const { createCipheriv , } = require( "node:crypto" );
const { Buffer } = require( "node:buffer" );

const SECRET = "DRAMA"

const Sha256 = createHash( "sha256" );
Sha256.update( SECRET );
const KEY = Sha256.digest();

const DATA = Buffer.from( "HELLO" , "utf8" );

function Decrypt( CIPHERTEXT , KEY ) {
    const NONCE = CIPHERTEXT.slice( 0 , 16 );
    const NONCEBuffer = Buffer.from( NONCE , "base64url" );
    const TAG = CIPHERTEXT.slice( -22 , );
    const TAGBuffer = Buffer.from( TAG , "base64url" );
    const ENCDATA = CIPHERTEXT.slice( 16 , -22 );
    const ENCDATABuffer = Buffer.from( ENCDATA , "base64url" );
    const AES256GCM = createDecipheriv( "aes-256-gcm" , KEY , NONCEBuffer );
    AES256GCM.setAuthTag( TAGBuffer );
    let DATABuffer = AES256GCM.update( ENCDATABuffer , null , "utf8" );
    DATABuffer = DATABuffer + AES256GCM.final( "utf8" );
    return DATABuffer
}

function Encrypt( DATABuffer , KEY ) {
    const NONCEBuffer = randomBytes( 12 );
    const NONCE = NONCEBuffer.toString( "base64url" );
    const AES256GCM = createCipheriv( "aes-256-gcm" , KEY , NONCEBuffer );
    const CIPHERBuffer = Buffer.concat( [ AES256GCM.update( DATABuffer ), AES256GCM.final( ) , AES256GCM.getAuthTag( ) ] );
    let CIPHERTEXT = CIPHERBuffer.toString( "base64url" );
    CIPHERTEXT = NONCE + CIPHERTEXT ;
    return CIPHERTEXT
}


// const x = Decrypt( "jhRGujr7QMzw_d7x4EM2jhsqEoCv2wTYLw-O_kr6pBQIg" , KEY );
// const x = Encrypt( DATA , KEY );
// console.log( x )

File: aes-gcm.go
package main

import "crypto/sha256"
import "crypto/aes"
import "crypto/cipher"
import "crypto/rand"
import "encoding/base64"
import "fmt"

func Encrypt( DATABuffer []byte , KEY string ) string {
    KEYHASH := sha256.Sum256( []byte( KEY ) )
    AES256 , _ := aes.NewCipher( KEYHASH[ : ] )
    AES256GCM , _ := cipher.NewGCM( AES256 )
    NONCEBuffer := make( []byte , 12 )
    _ , _ = rand.Read( NONCEBuffer )
    NONCE := base64.RawURLEncoding.EncodeToString( NONCEBuffer )
    ENCDataBuffer := AES256GCM.Seal( nil , NONCEBuffer , DATABuffer , nil )
    ENCData := base64.RawURLEncoding.EncodeToString( ENCDataBuffer )
    return NONCE + ENCData
}

func Decrypt( ENCData string , KEY string ) []byte {
    KEYHASH := sha256.Sum256( []byte( KEY ) )
    AES256 , _ := aes.NewCipher( KEYHASH[ : ] )
    AES256GCM , _ := cipher.NewGCM( AES256 )
    NONCEBuffer , _ := base64.RawURLEncoding.DecodeString( ENCData[ 0 : 16 ] )
    ENCDataBuffer , _ := base64.RawURLEncoding.DecodeString( ENCData[ 16 : ] )
    DATABuffer , _ := AES256GCM.Open( nil , NONCEBuffer , ENCDataBuffer , nil )
    return DATABuffer
}

func main( ) {
    SECRET := "DRAMA"
    DATA := []byte( "HELLO" )
    EncryptedData := Encrypt( DATA , SECRET )
    // fmt.Println( EncryptedData )
    out := Decrypt( EncryptedData , SECRET )
    fmt.Printf( "%s\n" , out )
}
